### ISSUE TITLE
Consider exposing a Parser property in MessageDescriptor

### ISSUE THREAD

**Issue by jskeet** [google/protobuf#806]
Generated code knows about a parser for the message type - so when we're parsing JSON, it would make sense to use that, but we only have a descriptor.

If we exposed a `Parser` property (of some new *non-generic* interface type, implemented explicitly by `MessageParser<T>`) we could avoid reflection in JSON parsing.

Note that this would be a codegen and runtime change, so would break existing generated code - we would want to get it in before 1.0.

@jtattermusch @xfxyjwf @anandolee for comment. I can create a prototype PR if that would help visualize things.

**Comment by xfxyjwf** [comment]
How much does this affect the generated code size? Does it share the same parsing code used to parse binary data?

**Comment by jskeet** [comment]
> How much does this affect the generated code size?

Hardly at all. Passing one more argument to the method to create a `Descriptor`.

> Does it share the same parsing code used to parse binary data?

Yes - it's not changing anything about how parsing's done, it just makes a parser accessible via the `Descriptor`.

**Comment by xfxyjwf** [comment]
How does this work with pure dynamic messages? I.e., a descriptor may not be associated with a generated type.

In C++ and Java, there isn't a one-to-one mapping from descriptor to message type. You can create both generated message type and dynamic message type from the same descriptor and giving only the descriptor, there is no way to map that to a concrete message type. Is the MessageDescriptor in C# different from the descriptor in C++ and Java?

**Comment by jskeet** [comment]
We don't currently have dynamic message types, but when we do we could still return a suitable parser from the descriptor - which would return a `DynamicMessage`, which still implements `IMessage`. Our `MessageDescriptor` has a `GeneratedType` property, but the intention is that that will return `null` if there isn't a generated type associated with it.

**Comment by xfxyjwf** [comment]
Can a MessageDescriptor already associated with a GeneratedType be used to create DynamicMessage? Or users will need to create new MessageDescriptors in order to create DynamicMessage? And is the dynamic message support planed to be included in 1.0?

(sorry, seems I have too many questions to ask here...)

**Comment by jskeet** [comment]
Hmm... hadn't considered that situation, as I'd normally envisaged the descriptor just being loaded from a byte array. We should definitely allow dynamic messages to be generated from message descriptors though... not sure whether that would involve cloning the descriptor objects, or just remembering the context when we parse into such a message.

**Comment by jskeet** [comment]
(Meant to say this before.) As for dynamic message support for 1.0 - unsure. We definitely want to get all the proto3 features in first, most importantly JSON parsing including handling well-known types - Any being the obviously-tricky one.

I would rather not get dynamic message support at all than get the *wrong* support - but let's see what we can work out. If we have a good set of known use cases for dynamic messages, that would be useful for API checking.

**Comment by xfxyjwf** [comment]
What is the GeneratedType property used for? In Java,  a Descriptor does not have any field that refers to the concrete message type, and the same Descriptor object can be shared by generated message types and dynamic message types. It seems C#'s MessageDescriptor is bound to a concrete type (a generated type or a dynamic message type). This is a different design from Java/C++ and I guess we should think it through before committed to it (especially on how dynamic message can be supported).

Some related questions:

A MessageDescriptor may be associated with a dynamic message so maybe we should name the property as MessageType rather than GeneratedType?

In Java we can compare two messages' descriptors by identity to determine whether they are of the same proto type (a generated message and a dynamic message can be of the same proto type if they are constructed on the same descriptor). Some reflection based methods (like copyFrom) will use this to determine it should proceed or complain about type mismatch. These methods will work seamlessly for both generated type and dynamic type. How would this work with C#'s MessageDescriptor?

About the use case for dynamic message:

The most popular use case I can see in our internal code base involves proto dbs. Basically you get a proto db (either a global db containing all message types, or a local db that comes with the data file you are parsing), retrieve descriptors from the db, construct DynamicMessages to parse the data and do queries/modifications on the dynamic message. For example, a comand-line tool that accepts sql like queries to analyze data files containing protos. Another good example is a grpc client tool that can get descriptor info from any grpc server and do queries (the equivalence of our stubby command line tool). Though arguably these tools are probably better off to use C++ rather than other languages. Supporting dynamic message is something good to have but not a must in other languages in my opinion.

**Comment by jskeet** [comment]
The `GeneratedType` is helpful for creating accessors via reflection. It doesn't *have* to be exposed, but it's helpful at the moment. I can see how having a "message-type-unaware" descriptor could be useful in some ways, but there are distinct benefits in it "knowing" the relevant generated types as well.

The point about comparing message descriptors by identity only works if the `DynamicMessage` *has* been created with an existing descriptor - it wouldn't help if the descriptor for the dynamic message had been loaded from a global db, for example, because then it wouldn't be the same object as the generated type's descriptor.

The use case you describe for dynamic messages is exactly the one I was thinking of - and it's one where you wouldn't have the generated types available to create the dynamic messages from to start with. That use case can definitely be served by my current design, with an appropriate `DynamicMessage` type which is given descriptors just as raw data (including dependencies, of course).

So basically, I agree that the design isn't the same as the Java and C++ APIs, but I think it'll make life simpler in a bunch of real use cases, without harming the dynamic message use case we *really* care about. I would like to get that use case of `DynamicMessage` implemented by 1.0, but probably won't go into more obscure ones.

**Comment by jskeet** [comment]
I've weighed up the pros and cons of this, and looking at the "any" case, I think it's going to be significantly simpler if we go ahead and add the `Parser` property. This will also require a non-generic MessageParser base class for `MessageParser<T>`, but that's okay.

**Comment by jskeet** [comment]
This is now implemented.
